allensdk/test/brain_observatory/behavior/data_files/test_stimulus_file.py

```diff
@@ -0,0 +1,91 @@
+from typing import Tuple
+from pathlib import Path
+import pickle
+from unittest.mock import create_autospec
+
+import pytest
+
+from allensdk.internal.api import PostgresQueryMixin
+from allensdk.brain_observatory.behavior.data_files import StimulusFile
+
+
+@pytest.fixture
+def stimulus_file_fixture(request, tmp_path) -> Tuple[Path, dict]:
+    default_stim_pkl_data = {"a": 1, "b": 2, "c": 3}
+    stim_pkl_data = request.param.get("pkl_data", default_stim_pkl_data)
+    stim_pkl_filename = request.param.get("filename", "test_stimulus_file.pkl")
+
+    stim_pkl_path = tmp_path / stim_pkl_filename
+    with stim_pkl_path.open('wb') as f:
+        pickle.dump(stim_pkl_data, f)
+
+    return (stim_pkl_path, stim_pkl_data)
+
+
+@pytest.mark.parametrize("stimulus_file_fixture", [
+    ({"pkl_data": {"a": 42, "b": 7}}),
+    ({"pkl_data": {"slightly_more_complex": [1, 2, 3, 4]}})
+], indirect=["stimulus_file_fixture"])
+def test_stimulus_file_from_json(stimulus_file_fixture):
+    stim_pkl_path, stim_pkl_data = stimulus_file_fixture
+
+    # Basic test case
+    input_json_dict = {"behavior_stimulus_file": str(stim_pkl_path)}
+    stimulus_file = StimulusFile.from_json(input_json_dict)
+    assert stimulus_file.data == stim_pkl_data
+
+    # Now test caching by deleting the stimulus_file
+    stim_pkl_path.unlink()
+    stimulus_file_cached = StimulusFile.from_json(input_json_dict)
+    assert stimulus_file_cached.data == stim_pkl_data
+
+
+@pytest.mark.parametrize("stimulus_file_fixture, behavior_session_id", [
+    ({"pkl_data": {"a": 42, "b": 7}}, 12),
+    ({"pkl_data": {"slightly_more_complex": [1, 2, 3, 4]}}, 8)
+], indirect=["stimulus_file_fixture"])
+def test_stimulus_file_from_lims(stimulus_file_fixture, behavior_session_id):
+    stim_pkl_path, stim_pkl_data = stimulus_file_fixture
+
+    mock_db_conn = create_autospec(PostgresQueryMixin, instance=True)
+
+    # Basic test case
+    mock_db_conn.fetchone.return_value = str(stim_pkl_path)
+    stimulus_file = StimulusFile.from_lims(mock_db_conn, behavior_session_id)
+    assert stimulus_file.data == stim_pkl_data
+
+    # Now test caching by deleting stimulus_file and also asserting db
+    # `fetchone` called only once
+    stim_pkl_path.unlink()
+    stimfile_cached = StimulusFile.from_lims(mock_db_conn, behavior_session_id)
+    assert stimfile_cached.data == stim_pkl_data
+
+    # This query string has strict formatting requirements
+    # in order to pass Mock assert_called_once_with() so don't change it!
```

Comment:
an alternative would be to store the query as a global variable in `stimulus_file.py` so that it could be imported in the test, and, keeping 2 copies of the string, with formatting, in-sync would not be necessary.
If going that way, it wouldn't be an f-string, but a string with placeholder `{}`

Reply:
Good point, will make this change.